### PR TITLE
Certificate processing

### PR DESCRIPTION
**Pull request by jricher**
Allows processing of MTLS bound access tokens through the library. Extraction functions around in BaseEndpoint, but these could be moved. Currently headers are hardcoded but these should be configurable.

### CODE REVIEW

File: src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java [hunk 1]
```diff
@@ -95,14 +95,14 @@ public String extractAccessToken(String authorization, String accessTokenInReque
...
     public AccessTokenInfo validateAccessToken(AuthleteApi api, String accessToken) throws WebApplicationException
     {
-        return validateAccessToken(api, accessToken, null, null);
+        return validateAccessToken(api, accessToken, null, null, null);
     }
 
 
     /**
      * Validate an access token. This method is an alias of {@link
      * #validateAccessToken(AuthleteApi, String, String[], String)
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java]: `#validateAccessToken(AuthleteApi, String, String[], String)` to
`#validateAccessToken(AuthleteApi, String, String[], String, String)`

File: src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java [hunk 2]
```diff
@@ -95,14 +95,14 @@ public String extractAccessToken(String authorization, String accessTokenInReque
...
-     * validate}<code>(api, accessToken, requiredScopes, null)</code>.
+     * validate}<code>(api, accessToken, requiredScopes, null, null)</code>.
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java]: Sorry, the JavaDoc before the change was wrong.
`validate}` should be changed to `validateAccessToken}`

File: src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java [hunk 3]
```diff
@@ -129,7 +129,45 @@ public AccessTokenInfo validateAccessToken(AuthleteApi api, String accessToken)
...
     {
-        return validateAccessToken(api, accessToken, requiredScopes, null);
+        return validateAccessToken(api, accessToken, requiredScopes, null, null);
+    }
+
+
+    /**
+     * Validate an access token. This method is an alias of {@link
+     * #validateAccessToken(AuthleteApi, String, String[], String)
+     * validate}<code>(api, accessToken, requiredScopes, requiredSubject, null)</code>.
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java]: Sorry, the JavaDoc before the change was wrong.
`validate}` should be changed to `validateAccessToken}`

File: src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java [hunk 4]
```diff
@@ -129,7 +129,45 @@ public AccessTokenInfo validateAccessToken(AuthleteApi api, String accessToken)
...
             AuthleteApi api, String accessToken, String[] requiredScopes) throws WebApplicationException
     {
-        return validateAccessToken(api, accessToken, requiredScopes, null);
+        return validateAccessToken(api, accessToken, requiredScopes, null, null);
+    }
+
+
+    /**
+     * Validate an access token. This method is an alias of {@link
+     * #validateAccessToken(AuthleteApi, String, String[], String)
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/BaseResourceEndpoint.java]: `#validateAccessToken(AuthleteApi, String, String[], String)` to
`#validateAccessToken(AuthleteApi, String, String[], String, String)`

File: src/main/java/com/authlete/jaxrs/BaseTokenEndpoint.java [hunk 5]
```diff
@@ -70,21 +70,28 @@
      *
      * @param authorization
      *         The value of {@code Authorization} header of the token request.
+     *         
+     * @param clientCertificate
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/BaseTokenEndpoint.java]: It seems the `handle` method does not have this parameter.

File: src/main/java/com/authlete/jaxrs/HeaderClientCertificateExtractor.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * 
+ * Extracts the client certificate from headers defined by the {@code 
+ * clientCertificatePathHeaders} member list. The first element in the list is
```
Comment (TakahikoKawasaki) [inline src/main/java/com/authlete/jaxrs/HeaderClientCertificateExtractor.java]: `clientCertificatePathHeaders` to `clientCertificateChainHeaders`?

File: src/main/java/com/authlete/jaxrs/HttpsRequestClientCertificateExtractor.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                    sb.append("-----BEGIN CERTIFICATE-----\n");
+                    sb.append(base64.encode(certs[i].getEncoded()));
+                    sb.append("\n-----END CERTIFICATE-----\n");
+
+                    pemEncoded[i] = sb.toString();
+                    
+                }
+            } catch (CertificateEncodingException e)
+            {
+                // TODO What should be done with this error?
```
[the remark this answers is not in the document]
Comment (jricher) [reply src/main/java/com/authlete/jaxrs/HttpsRequestClientCertificateExtractor.java]: There is no `ApiException` class, or any `Exception` classes that I can see, within this project.

**Comment by jricher** [comment]
Javadoc has been updated, still need a mechanism for throwing an exception or otherwise handling malformed certificate inputs.

**Pull request merged**